fearless/ApplicationLayer/Services/WalletBalanceSubscriptionAdapter.swift

```diff
@@ -0,0 +1,470 @@
+    private weak var delegate: WalletBalanceFetcherHandler?
+
+    private lazy var accountInfosAdapters: [String: AccountInfoSubscriptionAdapter] = [:]
+    private lazy var accountInfos: [ChainAssetKey: AccountInfo?] = [:]
+    private lazy var chainAssets: [ChainAsset] = []
+    private lazy var metaAccounts: [MetaAccountModel] = []
+    private lazy var prices: [PriceData] = []
+
+    // MARK: - Constructor
+
+    init(
+        metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>,
+        priceLocalSubscriptionFactory: PriceProviderFactoryProtocol,
+        chainRepository: AnyDataProviderRepository<ChainModel>,
+        operationQueue: OperationQueue,
+        eventCenter: EventCenterProtocol,
+        logger: Logger
+    ) {
+        self.priceLocalSubscriptionFactory = priceLocalSubscriptionFactory
+        self.metaAccountRepository = metaAccountRepository
+        self.chainRepository = chainRepository
+        self.operationQueue = operationQueue
+        self.eventCenter = eventCenter
+        self.logger = logger
+        eventCenter.add(observer: self, dispatchIn: .global())
+    }
+
+    // MARK: - Public methods
+
+    func subsctibeWalletBalance(
+        walletId: String?,
+        deliverOn queue: DispatchQueue?,
+        handler: WalletBalanceFetcherHandler
+    ) {
+        deliverQueue = queue
+        delegate = handler
+
+        if let identifier = walletId {
+            fetchMetaAccount(by: identifier)
+        } else {
+            fetchAllMetaAccounts()
+        }
+    }
+
+    // MARK: - Private methods
+
+    private func buildBalance() {
+        guard
+            accountInfos.isNotEmpty,
+            prices.isNotEmpty
+        else {
+            return
+        }
+
+        let builderBlock = BlockOperation {
+            let walletBalances = self.walletBalanceBuilder.buildBalance(
+                for: self.accountInfos,
+                self.metaAccounts,
+                self.chainAssets,
+                self.prices
+            )
+
+            guard let walletBalances = walletBalances else {
+                return
+            }
+
+            self.deliverOnCompletion(.success(walletBalances))
+        }
+
+        operationQueue.addOperation(builderBlock)
+    }
+
+    private func handle(_ wallets: [MetaAccountModel], _ chains: [ChainModel]) {
+        let chainsAssets = chains.map(\.chainAssets).reduce([], +)
+        chainAssets = chainsAssets
+        metaAccounts = wallets
+
+        subscribeToAccountInfo(for: wallets, chainsAssets)
+        subscribeToPrices(for: chains)
+    }
+
+    private func fetchMetaAccount(by identifier: String) {
+        let metaAccountOperation = metaAccountRepository.fetchOperation(
+            by: identifier,
+            options: RepositoryFetchOptions()
+        )
+        let chainsOperation = fetchChainsOperation()
+
+        metaAccountOperation.addDependency(chainsOperation)
+
+        metaAccountOperation.completionBlock = { [weak self] in
```

Comment:
it's okay and worsk but actually metaAccountOperation doesn't need result of chainsOperation to execute and it's incorrect by logic to say that metaAccountOperation depends on chainsOperation. it will be more beauty if you create another one operation called 'mergeOperation' with 2 dependencies - chainsOperation and metaAccountOperation.